stack.yaml

```diff
@@ -0,0 +1,71 @@
+# This file was automatically generated by 'stack init'
+#
+# Some commonly used options have been documented as comments in this file.
+# For advanced use and comprehensive documentation of the format, please see:
+# https://docs.haskellstack.org/en/stable/yaml_configuration/
+
+# Resolver to choose a 'specific' stackage snapshot or a compiler version.
+# A snapshot resolver dictates the compiler version and the set of packages
+# to be used for project dependencies. For example:
+#
+# resolver: lts-3.5
+# resolver: nightly-2015-09-21
+# resolver: ghc-7.10.2
+# resolver: ghcjs-0.1.0_ghc-7.10.2
+#
+# The location of a snapshot can be provided as a file or url. Stack assumes
+# a snapshot provided as a file might change, whereas a url resource does not.
+#
+# resolver: ./custom-snapshot.yaml
+# resolver: https://example.com/snapshots/2018-01-01.yaml
+resolver: lts-12.9
```

Comment:
The resolver is quite obscure to me: if I understand this correctly, it is just a repository version. What I did is searching for packages on hackage, then pick the latest version (as one should) and finally adding them here under `extra-deps`. Is that OK or is the preferred approach another one?

Reply:
A resolver refers to a snapshot (I think there is an ongoing discussion whether this should be renamed). A snapshot is a set of packages pinned at a specific version. A snapshot also has a specific GHC version (in our case 8.4.3).

So, knowing that, it is common (I'd dare say that's the whole point), that a snapshot does not have the latest version of packages. It has a set of packages that are known to work (= are stable). They undergo testing before being accepted into a LTS snapshot.

Now, "picking the latest version" isn't what I've been doing in Haskell. In my experience, unless you need a specific feature that has recently been added, sticking to the package versions provided by the snapshot is fine/encouraged.

When a package is not available in a snapshot, or at a lower version, you can add it to `extra-deps` essentially modifying the snapshot (does that make sense? like an override). And this should be the only reason you ever add a package there.

You can check out which packages are available and in which versions by looking at the snapshot on stackage:

https://www.stackage.org/lts-12.9
https://www.stackage.org/lts-12.9/package/yaml-0.8.32

So instead of going to Hackage and taking the newest version, I'd suggest you go to Stackage. I've asked Twitter if it is even necessary to add version constraints. I assume it is, but never do it for my personal projects.

</wall-of-text>

Did a bit of testing and turns out all you need to add is:

```yaml
extra-deps:
- yaml-0.10.1.1
```

as well as (in the package.yaml dependencies)

```yaml
- containers >= 0.5.11.0
```

and it compiles.

Reply:
All clear. One question remains though: is the resolver immutable? If it is mutable (updates, flow in), then using version constraints is probably wise anyway

Reply:
Good question, I had no idea. Apparently major versions are locked. Minor versions (and, therefore, hopefully without breaking API changes) get updates: https://github.com/commercialhaskell/stackage#frequently-asked-questions

TIL

Reply:
As discussed on the twitters, I will use the caret operator for all major versions (where `0.x.y.z` has `x` as major in hackage/stackage)